umoci: add "raw unpack"

flx42 opened the pull request:
This is pretty basic, and duplicates quite some code with the existing `unpack` code, that's why I've tagged this as WIP, to get the ball going and discuss our options.

I would like this feature for cases where we only care about getting the final rootfs. Computation of the mtree file can take a significant amount of time on large images (because of all the checksums that need to be computed).

thoughts @cyphar?

Closes #23

cyphar commented on cmd/umoci/raw-unpack.go:
```diff
@@ -0,0 +1,171 @@
+			return errors.Wrapf(err, "failure parsing --gid-map %s: %s", gidmap)
+		}
+		meta.MapOptions.GIDMappings = append(meta.MapOptions.GIDMappings, idMap)
+	}
```
Since this code is now used by `raw unpack` (and will be used by `insert` in #237) this should be moved to a separate function.

cyphar commented:
I'm totally fine with having a new `raw unpack` command -- as you said it's something that quite a few people have asked for. The current status looks fine (though there's a bit of duplication we should fix). Not to mention this obviously needs tests (you could probably just copy the `unpack` tests and turn them into `raw unpack` tests -- maybe there's a way to refactor it) and documentation.

cyphar commented on oci/layer/unpack.go:
```diff
@@ -120,19 +117,50 @@ func UnpackManifest(ctx context.Context, engine cas.Engine, bundle string, manif
 
+	defer func() {
+		if err != nil {
+			fsEval := fseval.DefaultFsEval
```
AFAICS this logic is wrong (if you intended to not use `unpriv.RemoveAll` you could just do `os.RemoveAll`). You need to do something like in `UnpackRootfs`:

flx42 replied:
You're right, I think I was misunderstanding why this particular is cleanup is needed.
Are you ok with a little duplication between the two function, or should we put that in a helper function `removeAll`?

cyphar replied:
I'd just duplicate it, personally.

flx42 commented:
Ok, so fixed the deferred removal. Added a few tests. But it depends on #240 
Next would be writing the doc, if we agree on the rest.

flx42 commented:
Rebased on top of current `master`, added the man page.
Do we need more tests?

Also, it keeps the old behavior of failing if the target rootfs folder already exists, are you OK with that?

cyphar commented:
This needs a rebase (again) -- sorry about that. The tests look fine to me.

> Also, it keeps the old behavior of failing if the target rootfs folder already exists, are you OK with that?

Yes I'm okay with that. We can always change it later.

cyphar commented:
LGTM. Will merge after test success.

flx42 commented:
Sorry, with DockerCon last week I didn't have time to do the rebase.

cyphar commented:
No problem. It was trivial to rebaseand it's all been merged. I will do a release with these features once I clean up the `umoci insert` stuff (it **needs** opaque whiteout support, and documentation, as well as a little bit of refactoring).
